### PR TITLE
COMMERCE-1844 Auto import regions of main countries out of box

### PR DESCRIPTION
**Pull request by riccardo-alberti**
Added auto import regions for: ESP, CHN, CUW, FIN, FRA, DEU, ISR, HUN, JPN, BRA, GBR, IND, CAN, KOR, RUS, AUS, MEX, IDN, NLD, TUR, CHE, SAU, ARG, POL

### CODE REVIEW

**Comment by riccardo-alberti** [comment]
ci:test

**Comment by ethanbustad** [comment]
Hey @riccardo-alberti, you'll need to follow these steps before CI will respond to test commands: https://grow.liferay.com/people/Pull+Request+Tester+for+Liferay+Developers.

**Review by ethanbustad — APPROVED** [review]
Nice job, @riccardo-alberti ! Looks great. We could consider reducing duplication in the `CommerceRegionsStarter`s, but they're already slim enough that this isn't particularly important in my mind. Nicely done!

**Comment by ethanbustad** [comment]
And congratulations on your first Commerce PR! 🎉 🚀

**Comment by marco-leo** [comment]
@riccardo-alberti @ethanbustad  I would like to create a base class or a Util, as looks like most of the json parsing code is duplicated.  What do you think?

**Comment by ethanbustad** [comment]
> @riccardo-alberti @ethanbustad I would like to create a base class or a Util, as looks like most of the json parsing code is duplicated. What do you think?

Yeah a base class was my first thought as well -- it just requires a bit of boilerplate still that way. But I think it'd be great to reduce duplication however we can.

**Comment by riccardo-alberti** [comment]
> > @riccardo-alberti @ethanbustad I would like to create a base class or a Util, as looks like most of the json parsing code is duplicated. What do you think?
> 
> Yeah a base class was my first thought as well -- it just requires a bit of boilerplate still that way. But I think it'd be great to reduce duplication however we can.

Yep agreed... i'll work on this

**Comment by marco-leo** [comment]
ci:test

File: commerce-service/src/main/java/com/liferay/commerce/internal/starter/ArgentinaCommerceRegionsStarter.java [hunk 1]
```diff
@@ -36,52 +31,22 @@
...
-				commerceCountry.getCommerceCountryId(), name, code, priority,
-				true, serviceContext);
-		}
+		start(
+			_commerceCountryLocalService, _commerceRegionLocalService,
+			_jsonFactory, serviceContext, ARGENTINA_NUMERIC_ISO_CODE,
+			layoutsPath);
 	}
 
+	private static final String layoutsPath =
```
Comment (ethanbustad) [inline commerce-service/src/main/java/com/liferay/commerce/internal/starter/ArgentinaCommerceRegionsStarter.java]: @riccardo-alberti these `layoutsPath` fields should be `_LAYOUTS_PATH` -- that's our naming convention for private constants.

File: commerce-service/src/main/java/com/liferay/commerce/internal/starter/CommerceRegionsStarterBase.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import com.liferay.portal.kernel.json.JSONArray;
+import com.liferay.portal.kernel.json.JSONFactory;
+import com.liferay.portal.kernel.json.JSONObject;
+import com.liferay.portal.kernel.service.ServiceContext;
+import com.liferay.portal.kernel.util.StringUtil;
+
+/**
+ * @author Riccardo Alberti
+ */
+public class CommerceRegionsStarterBase {
```
Comment (ethanbustad) [inline commerce-service/src/main/java/com/liferay/commerce/internal/starter/CommerceRegionsStarterBase.java]: Should be called `BaseCommerceRegionsStarter`, and in my opinion it should be `abstract` and implement `CommerceRegionsStarter`, to make it clearer how this class is intended to be used.

File: commerce-service/src/main/java/com/liferay/commerce/internal/starter/CommerceRegionsStarterBase.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+		if (commerceCountry == null) {
+			return;
+		}
+
+		parseAndAddCommerceRegions(
+			commerceCountry, serviceContext, layoutsPath);
+	}
+
+	private CommerceCountry getCommerceCountry(long companyId, int isoCode)
```
Comment (ethanbustad) [inline commerce-service/src/main/java/com/liferay/commerce/internal/starter/CommerceRegionsStarterBase.java]: `private` method names should start with an underscore -- i.e. `_getCommerceCountry`. Same for the other private methods below.

**Review by ethanbustad — CHANGES_REQUESTED** [review]

**Comment by ethanbustad** [comment]
ci:test

**Comment by marco-leo** [comment]
ci:test

File: commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	implements CommerceRegionsStarter {
+
+	protected void start(
+			CommerceCountryLocalService commerceCountryLocalService,
+			CommerceRegionLocalService commerceRegionLocalService,
+			JSONFactory jsonFactory, ServiceContext serviceContext, int isoCode,
+			String layoutsPath)
+		throws Exception {
+
+		_init(
```
Comment (marco-leo) [inline commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java]: @riccardo-alberti no need to init

File: commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+import com.liferay.portal.kernel.service.ServiceContext;
+import com.liferay.portal.kernel.util.StringUtil;
+
+/**
+ * @author Riccardo Alberti
+ */
+public abstract class BaseCommerceRegionsStarter
+	implements CommerceRegionsStarter {
+
+	protected void start(
```
Comment (marco-leo) [inline commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java]: @riccardo-alberti never pass refernce in method, use constructor

File: commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+			String name = jsonObject.getString("name");
+			double priority = jsonObject.getDouble("priority");
+
+			_commerceRegionLocalService.addCommerceRegion(
+				commerceCountry.getCommerceCountryId(), name, code, priority,
+				true, serviceContext);
+		}
+	}
+
+	private CommerceCountryLocalService _commerceCountryLocalService;
```
Comment (marco-leo) [inline commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java]: @riccardo-alberti maybe is better to use -dsannotations-options: inherit

File: commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+		if (commerceCountry == null) {
+			return;
+		}
+
+		_parseAndAddCommerceRegions(
+			commerceCountry, serviceContext, layoutsPath);
+	}
+
+	private CommerceCountry _getCommerceCountry(long companyId, int isoCode)
```
Comment (marco-leo) [inline commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java]: @riccardo-alberti no need to wrap one 1 line in one method

File: commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+			jsonFactory);
+
+		CommerceCountry commerceCountry = _getCommerceCountry(
+			serviceContext.getCompanyId(), isoCode);
+
+		if (commerceCountry == null) {
+			return;
+		}
+
+		_parseAndAddCommerceRegions(
```
Comment (marco-leo) [inline commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java]: _importCommerceRegions

File: commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java (hunk 5 above)
Comment (marco-leo) [inline commerce-service/src/main/java/com/liferay/commerce/internal/starter/BaseCommerceRegionsStarter.java]: The start method will be public and will implement th eone from the interface

File: commerce-service/src/main/java/com/liferay/commerce/internal/starter/BrazilCommerceRegionsStarter.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+	@Override
+	public void start(ServiceContext serviceContext) throws Exception {
+		start(
+			_commerceCountryLocalService, _commerceRegionLocalService,
+			_jsonFactory, serviceContext, BRAZIL_NUMERIC_ISO_CODE,
+			_LAYOUTS_PATH);
+	}
+
+	private static final String _LAYOUTS_PATH =
```
Comment (marco-leo) [inline commerce-service/src/main/java/com/liferay/commerce/internal/starter/BrazilCommerceRegionsStarter.java]: @riccardo-alberti  the base class will have an abstact method getFilePath, thos classes willovveride just that method

**Review by marco-leo — CHANGES_REQUESTED** [review]

**Comment by riccardo-alberti** [comment]
ci:test

**Comment by ethanbustad** [comment]
ci:test

**Review by ethanbustad — APPROVED** [review]
Beautiful.

**Comment by marco-leo** [comment]
@riccardo-alberti Good job and congratulation on your first contribution to Liferay Commerce!

**Pull request merged**